src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -343,75 +347,51 @@ generateX x
              $ hoistEither $ errorNoSuggestions (ErrorFunctionWrongArgs a x fErr [])
            annotate Map.empty resT $ \a' -> Prim a' p
 
-    -- Cases require all their alternatives to have the same type, as well as the scrutinee and the alternatives to have
-    -- the same temporality.
-    -- This may be more restrictive than necessary, since lets allow different temporalities, but hopefully will not be an issue.
+    -- Cases require:
     --
-    -- TODO: cases with nested constructors must be converted to multiple nested cases.
-    -- For example:
-    -- > case x
-    -- >  | Some (Some v, None) -> v
-    -- >  | Some (None, Some v) -> v
-    -- >  | _                   -> 0
-    -- > end
-    --
-    -- this should be flattened to
-    -- > case x
-    -- > | Some y
-    -- > -> case y
-    -- >    | (l,r)
-    -- >    -> case l
-    -- >       | Some v
-    -- >       -> case r
-    -- >          | Some _
-    -- >          -> 0
-    -- >          | None
-    -- >          -> v
-    -- >          end
-    -- >       | None
-    -- >       -> case r
-    -- >          | Some v
-    -- >          -> v
-    -- >          | None
-    -- >          -> 0
-    -- >          end
-    -- >       end
-    -- >    end
-    -- > | None
-    -- > -> 0
+    --  1. Alternatives to have "join-able" types, i.e. no mixing of aggregates and elements.
+    --  2. The return temporality is the join of the scrutinee with the alternatives.
+    --  3. The type of the scrutinee is compatible with all patterns.
     --
     Case a scrut pats
      -> do (scrut', sub) <- generateX scrut
-           (pats', subs) <- unzip <$> mapM (generateP $ annResult $ annotOfExp scrut') pats
-
-           resT <- case pats' of
-                    [] -> hoistEither
-                        $ errorNoSuggestions
-                        $ ErrorEmptyCase a x
-                    ((_,alt):_)
-                       -> return $ annResult $ annotOfExp alt
-
-           mapM_ (\(_,alt) -> require a $ CEquals resT $ annResult $ annotOfExp alt) pats'
-
-           let tmpx = getTemporalityOrPure $ annResult $ annotOfExp scrut'
-           let tmpq = getTemporalityOrPure $ resT
-           require a $ CEquals tmpx tmpq
-           let t' = resT -- canonT $ Temporality retTmp resT
 
+           -- Destruct the scrutinee type into the base type
+           -- and the temporality (defaulting to Pure).
+           let scrutT  = annResult $ annotOfExp scrut'
+           scrutTy    <- TypeVar <$> fresh
+           scrutTm    <- TypeVar <$> fresh
+           require a $ CExtractTemporality scrutTm scrutTy scrutT
```

Comment:
That gives you the temporality of the scrutinee (or Pure). But I wanted to do was adding the unification variable `scrutTm` into the context, so that it will resolve to the right temporality when it discharges the joins.